packages/api/src/middlewares/headers.ts

```diff
@@ -0,0 +1,24 @@
+import type { PieceContext } from '@sapphire/framework';
+import { METHODS } from 'http';
+import type { ApiRequest } from '../lib/structures/api/ApiRequest';
+import type { ApiResponse } from '../lib/structures/api/ApiResponse';
+import { Middleware } from '../lib/structures/Middleware';
+
+export class PluginMiddleware extends Middleware {
+	private readonly origin: string;
+	private readonly methods: string = METHODS.join(', ');
+
+	public constructor(context: PieceContext) {
+		super(context, { priority: 10 });
+		this.origin = this.client.options.api?.origin ?? '*';
+	}
+
+	public run(request: ApiRequest, response: ApiResponse) {
+		response.setHeader('Date', new Date().toUTCString());
+		response.setHeader('Access-Control-Allow-Credentials', 'true');
+		response.setHeader('Access-Control-Allow-Origin', this.origin);
+		response.setHeader('Access-Control-Allow-Headers', 'Authorization, User-Agent, Content-Type');
```

Comment:
What if we may want to include extra headers? Can't we get something like origin?

Reply:
origin is the line 19.

This defines the CORS headers, though.

Reply:
I asked if we can get something LIKE `origin`, ONLY for `allowedHeaders`, but kyra doesn't read the whole sentence

Reply:
To be honest, I defined `origin` in NGINX many years ago. But yeah, I can implement that.

Reply:
Implemented.

```diff
diff --git a/packages/api/src/lib/Api.ts b/packages/api/src/lib/Api.ts
index 28d3ee6..1239ed3 100644
--- a/packages/api/src/lib/Api.ts
+++ b/packages/api/src/lib/Api.ts
@@ -44,6 +44,8 @@ export interface ApiOptions {
 	 */
 	origin?: string;
 
+	/**
+	 * @since 1.0.0
+	 * @default []
+	 */
+	extraHeaders?: [string, string][];
+
 	/**
 	 * @since 1.0.0
 	 * @default {}
diff --git a/packages/api/src/middlewares/headers.ts b/packages/api/src/middlewares/headers.ts
index e173fb4..59f56d1 100644
--- a/packages/api/src/middlewares/headers.ts
+++ b/packages/api/src/middlewares/headers.ts
@@ -7,10 +7,12 @@ import { Middleware } from '../lib/structures/Middleware';
 export class PluginMiddleware extends Middleware {
 	private readonly origin: string;
 	private readonly methods: string = METHODS.join(', ');
+	private readonly extraHeaders: readonly [string, string][];
 
 	public constructor(context: PieceContext) {
 		super(context, { priority: 10 });
 		this.origin = this.client.options.api?.origin ?? '*';
+		this.extraHeaders = this.client.options.api?.extraHeaders ?? [];
 	}
 
 	public run(request: ApiRequest, response: ApiResponse) {
@@ -19,6 +21,13 @@ export class PluginMiddleware extends Middleware {
 		response.setHeader('Access-Control-Allow-Origin', this.origin);
 		response.setHeader('Access-Control-Allow-Headers', 'Authorization, User-Agent, Content-Type');
 		response.setHeader('Access-Control-Allow-Methods', this.methods);
+		for (const [name, value] of this.extraHeaders) response.setHeader(name, value);
+
+		// RFC 7231 4.3.7:
+		// > This method allows a client to determine the options and/or requirements associated with a
+		// > resource, or the capabilities of a server, without implying a resource action.
+		//
+		// Due to this method's nature, it is recommended to end the request after setting pre-flight CORS headers.
 		if (request.method === 'OPTIONS') response.end();
 	}
 }
```